### ISSUE TITLE
mapping a set of puts is broken

### ISSUE THREAD

**Issue by sanoel** [OADA/client#15]
```
Promise.map(['a', 'b', 'c'], async (i) => { 
    await CONNECTION.put({
        path: '/bookmarks/test/'+i,
        tree: {
            bookmarks: {_type: '...', _rev: '...', 
                test: {_type: '...', _rev: '...', 
                    '*': { _type: '...', _rev: '...'} 
                 }
             }
        },
        data: {foo: 'bar'}
    })
})
```

The result:
Only 'a' gets created as a resource under /bookmarks/test. The remainder are just plain objects. 

**The temporary fix to just get a tree created properly is to use Promise.each. However, because we're not _actually_ protected against concurrent writes, this is only a solution for a single client trying to get a tree setup.** 

I spent some time going through the code and it seems a bit difficult to fix the way iteration is setup. This code is very clean and works well for single clients/non-concurrent writes. I like that it creates the entire "new" tree of data outside of the bookmarks tree then one final link joins this to the bookmarks tree. However...this is also its downfall: if when it reaches that final link into the main bookmarks graph it finds a link already there, the whole thing must essentially be thrown out as any common children resources are duplicates. 

The solution unfortunately entails a bit more complexity. It requires that we first find the highest existing resource and work down from there (as originally implemented in oada-cache). Then, we can actually perform the proper if-matches as we create those links from parent to child.

**More detailed problem walkthrough:**
-----------------
Because of the way linkObj is setup on the iteration prior to when it gets written/put, the if-match check will retry the _current_ iteration, but it will still use the wrong linkObj setup in the previous iteration, clobbering 'test' resources. 

In the example above, the first iteration creates the three /bookmarks/test/* resources, for example /bookmarks/test/a. Also prepared on this iteration is a linkObj of {_id: 'resources/a'} to be used on the _next_ iteration, so that the parent can point to it.

On the second iteration, /bookmarks/test does not yet exist as a resource and so it is created once for each item mapped over, (e.g. resources/testa, resources/testb, resources/testc all get created). The body of those new resources is constructed from the child path and the linkObj, e.g., {a: {_id: 'resources/a'}}. The linkObj prepared for the next iteration is {_id: resources/testa}.

On iteration three, /bookmarks exists. Because bookmarks exists, a different code path executes the following put using the linkObj and an if-match on the bookmarks rev retrieved when we checked that it existed, e.g.,: {path: /bookmarks/test, body: {_id: resources/testa}, rev: 2}. The if-match check does its job and makes 'b' and 'c' retry _this current iteration_ (because it was mapped, they keep pace with one another), but even when the correct rev is retrieved on /bookmarks, the put body (linkObj) is still wrong.

**Comment by abalmos** [comment]
@tomohiroarakawa Could you quickly indicate if you were trying to do it the `oada-cache` way or if you had another idea which may just have a bug?

A "quick" solution may be just restarting the entire tree put if there is a if-match error at any point?

**Comment by tomohiroarakawa** [comment]
The way oada/client creates a tree is different from the oada-cache way. It begins with creating a leaf resource (without checking the highest existing resource) and recursively check and create resources along the way to the root.

**Comment by sanoel** [comment]
I now realize you were in fact retrying all over again in the event of a 412. I initially misinterpreted the code as retrying "this current iteration". However, it looks like `continue` subtracts one from `i` and so it doesn't quite start over again. I think if we change `i = pathArray.length-1` to `i=pathArray.length`, it fixes it.

**Comment by tomohiroarakawa** [comment]
I will try to look into this issue later today. Thanks for the bug report.

**Comment by tomohiroarakawa** [comment]
@sanoel You're right, that was the bug. I merged your fix to `master`, added a new unit test case for concurrent PUTs, and published `oada/client` v2.3.1 to npm. I also adjusted the parameters of the exponential backoff, so the conflict resolution after the if-match error should be faster now.
